src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,1389 @@
+                // (11,20): error CS0029: Cannot implicitly convert type 'int' to 'System.Threading.Tasks.Task<int>'
+                //             return 1;
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "1").WithArguments("int", "System.Threading.Tasks.Task<int>").WithLocation(11, 20)
+                );
+        }
+
+        [Fact]
+        public void TestMissingIDisposable()
+        {
+            string source = @"
+class C
+{
+    int M()
+    {
+        using (new C())
+        {
+        }
+        using (var x = new C())
+        {
+            return 1;
+        }
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source);
+            comp.MakeTypeMissing(SpecialType.System_IDisposable);
+            comp.VerifyEmitDiagnostics(
+                // (6,9): error CS0518: Predefined type 'System.IDisposable' is not defined or imported
+                //         using (new C())
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "using").WithArguments("System.IDisposable").WithLocation(6, 9),
+                // (6,16): error CS1674: 'C': type used in a using statement must be implicitly convertible to 'System.IDisposable'
+                //         using (new C())
+                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "new C()").WithArguments("C").WithLocation(6, 16),
+                // (9,9): error CS0518: Predefined type 'System.IDisposable' is not defined or imported
+                //         using (var x = new C())
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "using").WithArguments("System.IDisposable").WithLocation(9, 9),
+                // (9,16): error CS1674: 'C': type used in a using statement must be implicitly convertible to 'System.IDisposable'
+                //         using (var x = new C())
+                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "var x = new C()").WithArguments("C").WithLocation(9, 16)
+                );
+        }
+
+        [Fact]
+        public void TestMissingDisposeAsync()
+        {
+            string source = @"
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        // missing DisposeAsync
+    }
+}
+class C : System.IAsyncDisposable
+{
+    async System.Threading.Tasks.Task<int> M()
+    {
+        using await (new C()) { }
+        using await (var x = new C()) { return 1; }
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source);
+            comp.VerifyDiagnostics();
+            comp.VerifyEmitDiagnostics(
+                // (13,9): error CS0656: Missing compiler required member 'System.IAsyncDisposable.DisposeAsync'
+                //         using await (new C()) { }
+                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "using await (new C()) { }").WithArguments("System.IAsyncDisposable", "DisposeAsync").WithLocation(13, 9),
```

Comment:
>using await (new C()) { } [](start = 67, length = 25)

It feels like we should narrow location of this error.

Reply:
No response for this comment.

---
In reply to: [162707185](https://github.com/dotnet/roslyn/pull/23961#discussion_r162707185) [](ancestors = 162707185)